miniwin/src/d3drm/backends/sdl3gpu/renderer.cpp

```diff
@@ -0,0 +1,381 @@
+	}
+
+	const SDL_GPUShaderCreateInfo* fragmentCreateInfo =
+		GetFragmentShaderCode(FragmentShaderId::SolidColor, SDL_GetGPUShaderFormats(device));
+	if (!fragmentCreateInfo) {
+		return nullptr;
+	}
+	SDL_GPUShader* fragmentShader = SDL_CreateGPUShader(device, fragmentCreateInfo);
+	if (!fragmentShader) {
+		return nullptr;
+	}
+
+	SDL_GPUVertexBufferDescription vertexBufferDescs[1] = {};
+	vertexBufferDescs[0].slot = 0;
+	vertexBufferDescs[0].pitch = sizeof(PositionColorVertex);
+	vertexBufferDescs[0].input_rate = SDL_GPU_VERTEXINPUTRATE_VERTEX;
+	vertexBufferDescs[0].instance_step_rate = 0;
+
+	SDL_GPUVertexAttribute vertexAttrs[3] = {};
+	vertexAttrs[0].location = 0;
+	vertexAttrs[0].buffer_slot = 0;
+	vertexAttrs[0].format = SDL_GPU_VERTEXELEMENTFORMAT_FLOAT3;
+	vertexAttrs[0].offset = 0;
+
+	vertexAttrs[1].location = 1;
+	vertexAttrs[1].buffer_slot = 0;
+	vertexAttrs[1].format = SDL_GPU_VERTEXELEMENTFORMAT_FLOAT3;
+	vertexAttrs[1].offset = sizeof(float) * 3;
+
+	vertexAttrs[2].location = 2;
+	vertexAttrs[2].buffer_slot = 0;
+	vertexAttrs[2].format = SDL_GPU_VERTEXELEMENTFORMAT_UBYTE4_NORM;
+	vertexAttrs[2].offset = sizeof(float) * 6;
+
+	SDL_GPUVertexInputState vertexInputState = {};
+	vertexInputState.vertex_buffer_descriptions = vertexBufferDescs;
+	vertexInputState.num_vertex_buffers = SDL_arraysize(vertexBufferDescs);
+	vertexInputState.vertex_attributes = vertexAttrs;
+	vertexInputState.num_vertex_attributes = SDL_arraysize(vertexAttrs);
+
+	SDL_GPUColorTargetDescription colorTargets = {};
+	colorTargets.format = SDL_GPU_TEXTUREFORMAT_R8G8B8A8_UNORM;
+
+	SDL_GPUGraphicsPipelineCreateInfo pipelineCreateInfo = {};
+	pipelineCreateInfo.vertex_shader = vertexShader;
+	pipelineCreateInfo.fragment_shader = fragmentShader;
+	pipelineCreateInfo.vertex_input_state = vertexInputState;
+	pipelineCreateInfo.primitive_type = SDL_GPU_PRIMITIVETYPE_TRIANGLELIST;
+	pipelineCreateInfo.target_info.color_target_descriptions = &colorTargets;
+	pipelineCreateInfo.target_info.num_color_targets = 1;
+	pipelineCreateInfo.target_info.has_depth_stencil_target = true;
+	pipelineCreateInfo.target_info.depth_stencil_format = SDL_GPU_TEXTUREFORMAT_D16_UNORM;
+	pipelineCreateInfo.depth_stencil_state.enable_depth_test = true;
+	pipelineCreateInfo.depth_stencil_state.enable_depth_write = true;
+	pipelineCreateInfo.depth_stencil_state.enable_stencil_test = false;
+	pipelineCreateInfo.depth_stencil_state.compare_op = SDL_GPU_COMPAREOP_GREATER;
+	pipelineCreateInfo.depth_stencil_state.write_mask = 0xff;
+
+	SDL_GPUGraphicsPipeline* pipeline = SDL_CreateGPUGraphicsPipeline(device, &pipelineCreateInfo);
+	// Clean up shader resources
+	SDL_ReleaseGPUShader(device, vertexShader);
+	SDL_ReleaseGPUShader(device, fragmentShader);
+
+	return pipeline;
+}
+
+Direct3DRMRenderer* Direct3DRMSDL3GPURenderer::Create(DWORD width, DWORD height)
+{
+	SDL_GPUDevice* device = SDL_CreateGPUDevice(
+		SDL_GPU_SHADERFORMAT_SPIRV | SDL_GPU_SHADERFORMAT_DXIL | SDL_GPU_SHADERFORMAT_MSL,
```

Comment:
Backend selection now works via the original games capability detection. Each backed implements the D3DDevice interface and is probed for it's capabilities, and only a working and fitting render will be selected.

If the game is ported to systems with with platform specific rendering backends they can simply be added at the top of the reported backends and enabled for that platform.

You can simulate a system that isn't able to support SDL3 GPU by removing editing this line to only have a format that isn't supported by your system (`SDL_GPU_SHADERFORMAT_DXIL` for Mac and `SDL_GPU_SHADERFORMAT_MSL` for everyone else).

Reply:
I'd still like to be able to select a backend, if only for testing purposes.

Reply:
I think a good compromise would be to honor the environment variable if it is set, and fail if the selected backend isn't available. But I like that the game will fall back by default to software. We should make it used friendly.

Reply:
`SDL_GPU_DRIVER` is an SDL environment variable, and should not be used. Please use one in the `ISLE_*` namespace.

Reply:
Ok, I can implement that as well. Is it ok if I do so as a follow up tomorrow?